WIP on 'visitor' model for the SemaSYCL refactor

erichkeane opened the pull request:
This is my first run at how I think we can re-arch this to better share work.  Suggestions/patches welcome :)

I've implemented ONLY the arg-type diagnostics so far, but I intend to start looking at other parts when I get a chance.

Signed-off-by: Erich Keane <erich.keane@intel.com>

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1307,6 +1297,153 @@ CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
+    else if (ArgTy->isScalarType())
+      KF_FOR_EACH(handleScalarType);
+    else
+      llvm_unreachable("Unsupported kernel parameter type");
```
This almost definitely shouldn't be here.  I should add a "handleOtherType" that is only implemented by the diagnostics, which simply diagnoses with an error.  Everyone else can ignore it.

MrSidims commented:
Just adding myself to reviewers to not miss what is going on.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1307,6 +1297,153 @@ CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
+};
+
+// A type to Create and own the FunctionDecl for the kernel.
+class SYCLKernelHeader : public SyclKernelFieldHandler<SYCLKernelHeader> {
```
Why this class is needed at all?
What is kernel obj header?

erichkeane replied:
I'm not sure yet :)  I was brainstorming based on the current things we do.  The idea was to have this + the error-checking version replace BuildArgTys.  This is more just to build/own the declaration and much of the stuff that happens in ConstructOpenCLKernel.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1307,6 +1297,153 @@ CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
+// A base type that the SYCL OpenCL Kernel construction task uses to implement
+// individual tasks.
+template<typename Derived>
+class SyclKernelFieldHandler {
```
We also need one of these for the integration header generation.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1307,6 +1297,153 @@ CreateOpenCLKernelDeclaration(ASTContext &Context, StringRef Name,
+public:
+  // TODO: Should the default behavior DO anything?  Also, do these need any
+  // more params?
+  void handleSyclAccessorType(const FieldDecl *, QualType){}
+  void handleSyclSamplerType(const FieldDecl *, QualType){}
```
I could swear I saw a spot where we handled them separately (but I don't have the terminal open, but I want to say the Integration Header?).  I would think we can keep them split for the near future, and combine them if we find they aren't particularly useful to be separate.

Fznamznon replied:
Ah, right. The handling in integration header can be different for accessors and samplers, but It should be identical for generation of OpenCL kernel body.

erichkeane replied:
Ok, that matches with my understanding as well! I presume the OpenCL Kernel body/decl will likely implement theirs in a function that this just calls.  Maybe I'll make that part a bit easier when I get there.

erichkeane commented:
Did a pretty sizable amount of work this evening, and go tmyself to the point where I think I have a pretty good framework.  I still don't have the body of the SYCL Kernel being generated, but I think I've replaced all of buildArgTys.  I also added the type for Integration Headers, but didn't do any implementation for it.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+
+    Params.push_back(NewParam);
+
+    // Create the new type.
+    ArgTys.push_back(ArgTy);
+    FunctionProtoType::ExtProtoInfo Info(CC_OpenCLKernel);
+    QualType FuncType = Ctx.getFunctionType(Ctx.VoidTy, ArgTys, Info);
+
+    // Set type, note until the destructor of the owner of KernelObj is called
+    // (to set the parameters), we cannot access the parameters from KernelObj
+    // without memory problems.
+    KernelObj->setType(FuncType);
```
I think we should move this part into another function, because we re-create and re-set kernel function type for each new argument, this is strange and doesn't seem effective. Here we can leave only collecting of argument types in `ArgTys` array.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+  }
+
+  void handlePointerType(const FieldDecl *FD, QualType ArgTy) final {
+    // TODO: Can we document what the heck this is doing?!
```
Yep, that would be nice. But it's not so hard to understand what is happening here looking into the code.
This is just handling of USM pointers. I haven't seen any doc about implementation details of USM, only specs (https://github.com/intel/llvm/tree/sycl/sycl/doc/extensions/USM).
Basically USM allows to use raw pointers instead of buffers/accessors, but these pointers point to the specially allocated memory. For any pointer we add some argument with the same type but global address space, because OpenCL requires it.
The problem here that we don't know how to differentiate USM pointers from any others, so we stupidly consider them all as USM pointers. This can lead to some problems, I guess.

erichkeane replied:
I've read both the code and your post and don't have a good hold of it other than it seems to be just creating a pointer to the same type, except global.  Patches that do a good job explaining the details would be welcome :)

jbrodman replied:
Hi. Sorry.

At least when I did this, kernel args to OpenCL that are pointers had to be in the global address space.

I make no claims to having done the right thing, only that I did something that got pointers to work.

And yes - I don't see a way to disambiguate 'normal' C++ pointers from USM pointers since USM is very specifically trying to avoid modifying the type system.
Previously, the compiler was rejecting all programs with pointers, so things that should work wouldn't.
We had to make it so any pointer works, which means that legal things now work, but illegal things will crash.  I still view that as an improvement.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
 
 // Generates the OpenCL kernel using KernelCallerFunc (kernel caller
-// function) defined is SYCL headers.
+// function) defined is Sycl headers.
```
Why?

erichkeane replied:
I was being incredibly inconsistent with my naming at one point, so I ran a find/replace.  I guess it got a little out of hand :)

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+    FD->addAttr(ArtificialAttr::CreateImplicit(Context));
+  }
+
+  static FunctionDecl *initKernelObj(ASTContext &Ctx, StringRef Name,
+                                     SourceLocation Loc, bool IsInline) {
```
I don't get why it is called `initKernelObj`. Shouldn't it be something like `createKernelDecl`?
BTW, I don't think that OpenCL kernel somewhat can be inline. o_O

erichkeane replied:
The implicit inline logic came from the existing implementation.

Naming things is hard... 'initKernelObj' (with the variable named KernelObj) made sense at the time *shrug*.

Fznamznon replied:
I think because the name `KernelObj` is incorrect.
```
FunctionDecl *KernelObj;
```
I think this variable contains declaration (and definition after we generate the body) of OpenCL kernel. We have to rename it so it will be easy to understand that this is OpenCL kernel function declaration. Something like `KernelDecl`.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+    // Create descriptors for each accessor field in the class or struct
+    const auto *Wrapper = ArgTy->getAsCXXRecordDecl();
+    SyclKernelDeclBase Recurse(SemaRef, KernelObj, ArgTys, Params);
+    VisitKernelFields(Wrapper->fields(), Recurse);
```
The `Wrapper` here is not a SYCL kernel object. Seeing Visit**Kernel**Fields seems strange here. Maybe we probably need to rename `VisitKernelFields` to something like `visitRecordFields`.

erichkeane replied:
Sure.

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+  }
+};
+
+// A type that handles the accessor recursion and acts as a base for
+// SyclKernelDeclCreator. It doesn't 'own' anything other than the KernelObj
+// pointer and functionality required to add a param.
+class SyclKernelDeclBase : public SyclKernelFieldHandler<SyclKernelDeclBase> {
```
I don't really get why do we need a base and a child class for kernel decl creation.
Could you please elaborate a bit more?

Fznamznon commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1256,65 +1182,331 @@ static void populateIntHeader(SYCLIntegrationHeader &H, const StringRef Name,
+  (void)std::initializer_list<int> { (handlers.FUNC(Field, ArgTy), 0)... }
+
+  for (const auto &Field : Fields) {
+    QualType ArgTy = Field->getType();
```
I prefer call this thing not `ArgTy` but `KernelObjFieldTy` or something like this... Because I was confused by it, I was thinking that it is type of kernel argument for some time.

erichkeane replied:
Well, the idea is that in this function it is a 'generic' function, so all it cares about is the field.  Perhaps FieldTy?

Fznamznon replied:
Ok. Maybe we also need to rename function so the name of it won't contain `Kernel` at all. I've seen this function is used for some structs, not only SYCL kernel object.

erichkeane replied:
Yep, thats a good idea.  It originally ONLY did the kernel fields :) I discovered the struct recursion later.

erichkeane commented:
I'll do a bunch of the renaming/changes that @Fznamznon  suggested on this patch.

erichkeane commented:
Ok, pushed a couple more commits.  These changed the look of the visitor slightly to handle recursion and base types.

Additionally, it SHOULD properly handle accessors as both fields and base types all the way up the tree.  

We need tests for that, there are quite a few TODOs, we still need the body implementation, it would be nice to get Elizabeth's implementation of memcpy working alright (perhaps after body + a squash?), and it would be nice to get arrays working as well.

For now, I'm just quitting for the weekend, I'll pick this up on monday.

rdeodhar commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1001,311 +1002,580 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+  // class/field graph. Int Headers use this to calculate offset, most others
+  // don't have a need for these.
+
+  virtual void enterStruct(const CXXRecordDecl *, const FieldDecl *) {}
```
I don't know if leaveStruct will ever be needed. The offsets generated for the integration header are monotonically increasing. If the lambda object members are visited in ascending sequence, the offset value should be steadily increasing, thus it will never need bumping down.

I confess I don't fully understand the base class stuff either. The old code made no reference to base classes.

erichkeane replied:
I did the 'leaveStruct' so that we can 'undo' the offset, that way we don't have to store the entire 'stack' of parents.  Instead, we can store just a single offset, and add/subtract it as needed.  For example, if we are deep in a set of structures and see:

    struct S {  accessor a;    };
    struct U { /*other fields;*/S s1, s2; };

We would have to 'know' about U and the entire set of parents in order to calculate the offset.  By doing the 'leaveStruct' (see the Int Header implementation), each time you enter a field, you simply increase by your offset, then decrease by the same offset when we leave.  So in the above:

We're in 'U', offset is something like 95 (random number).
We enter 's1', which we know is at offset '40' in 'U', so we increase our offset by that much.
We leave 's1', we can calculate the offset as '40' in 'U' and just decrease it, getting us back to 95.
Then, we enter 's2', we can do the same.

Otherwise, we'd have to know 'U''s offset, and our current field/bases' offset as well.  I couldn't come up with a way of doing that which didn't store the stack of currently visited things.

erichkeane commented on clang/lib/Sema/SemaSYCL.cpp:
```diff
@@ -1001,311 +1002,580 @@ static ParamDesc makeParamDesc(const FieldDecl *Src, QualType Ty) {
+    else if (Util::isSyclStreamType(ItemTy))
+      (void)std::initializer_list<int>{
+          (handlers.handleSyclStreamType(Item, ItemTy), 0)...};
+    else if (ItemTy->isStructureOrClassType()) {
```
Hmm... I didn't see in the old code it recursing through arrays, are we intending to split them out as well?

In that case, we probably need to better handle the offsets.

erichkeane replied:
Issue captured here: https://github.com/intel/llvm/issues/1567

rdeodhar replied:
It is quite natural to recurse through structs to any level to discover accessors and the new visitor model does that.
The existing implementation does not deal with arrays at all so you won't see any code there.
My support for arrays accounted for arrays of accessors, i.e., the array type is one of the special sycl types. When such an array type was encountered it would "unroll" each element of the array and generate a separate descriptor for it. It did not try to handle arrays of other struct types that contained accessors (or arrays of accessors) as members, or more deeply nested members.
I suggest we support arrays of accessors only, not arrays of structs that may in turn contain accessors.
If so, the proposed addition to the visitor function will do the job.

erichkeane replied:
I think we'd have to support both, arrays of structs of accessors, or even arrays of structs of arrays of accessors.

It is a presumably fairly easy addition however, but I'll hold it off for another patch, we have a fairly sizable change coming from  @elizabethandrews  on how parameters are passed, so I'd like to hold off on arrays until after that.

bader commented:
Should we close this PR?
#1517 seems to replace this one.
